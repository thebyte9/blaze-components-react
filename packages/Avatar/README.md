## Description

Avatar is the graphical representation of the user or the user's alter ego or character.

## Usage

- Image

```js
const url = 'http://lorempixel.com/400/400/people/';

<Avatar url={url} modifier={Avatar.availableModifiers.med} />

<Avatar
  url={url}
  username="Ismael Haytam"
  modifier={Avatar.availableModifiers.small}
/>
```

- User Initials

```js
<Avatar username="Blaz 2" modifier={Avatar.availableModifiers.med} />

<Avatar
  username="Lorem Ipsum"
  modifier={Avatar.availableModifiers.xSmall}
/>
```

## API

##### Avatar can receive a number of `props` as follow:

| NAME     |  TYPE  | DEFAULT | Options             |
| :------- | :----: | :-----: | ------------------- |
| modifier | String |  empty  | med, small, x-small |
| url      | string |  empty  |
| username | String |    !    |
