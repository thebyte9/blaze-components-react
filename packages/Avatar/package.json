{
    "name": "@blaze-react/avatar",
    "version": "0.0.1",
    "description": "",
    "main": "index.js",
    "scripts": {
        "build": "./node_modules/.bin/babel index.js -d lib",
        "build-storybook": "build-storybook -c .storybook -o docs-build",
         "lint": "eslint ./ --ignore-path ../../.eslintignore",

        "test": "jest",
        "test:coverage": "jest --coverage",
        "pre-publish": "npm run build && npm run lint && npm run test"
    },
    "keywords": [],
    "author": "",
    "license": "ISC",
    "dependencies": {},
    "devDependencies": {
        "@babel/cli": "^7.4.4",
        "@babel/core": "^7.4.5"
    },
    "peerDependencies": {
        "react": "^16.8.6",
        "react-dom": "^16.8.6",
        "prop-types": "^15.7.2"
    },
    "babel": {
        "presets": [
            "@babel/preset-env",
            "@babel/preset-react"
        ]
    }
}