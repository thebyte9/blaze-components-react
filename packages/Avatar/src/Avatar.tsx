import buildClassNames from '../../Utils/src/buildClassNames';
import React, { useState } from "react";

type Tmodfier = "med" | "small" | "x-small";

interface IAvatarProps {
  modifier?: Tmodfier;
  url?: string;
  username?: string;
  attr?: Record<string, unknown>;
}

const Avatar =
  ({
    modifier,
    url,
    username,
    ...attr
  }: IAvatarProps): JSX.Element => {
    const [avatarUrl, setAvatar] = useState<string | undefined>(url);
    const [validUrl, setValidUrl] = useState<boolean>(false);

    const avatarClassName: string = buildClassNames("avatar", {
      [`avatar--${modifier}`]: !!modifier
    });

    const initials = !username
      ? "!"
      : username
        .split(" ")
        .map(subName => subName[0].toUpperCase())
        .join("")
        .substring(0, 2);

    if (url && !validUrl) {
      const imageData = new Image();
      imageData.src = url;

      imageData.addEventListener(
        "load",
        (): void => {
          setAvatar(url);
          setValidUrl(true);
        }
      );
    }

    return (
      <div className={avatarClassName}>
        {validUrl ? (
          <img src={avatarUrl} alt="avatar" {...attr} />
        ) : (
          <span>{initials}</span>
        )}
      </div>
    );
  }
  ;

const availableModifiers = {
  med: "med",
  small: "small",
  xSmall: "x-small"
};

Avatar.availableModifiers = availableModifiers;

Avatar.defaultProps = {
  modifier: "",
  url: "",
  username: ""
};

export default Avatar;
