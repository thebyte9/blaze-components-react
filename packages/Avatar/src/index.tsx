import React, { FunctionComponent, useState } from "react";
interface IAvatarProps {
  modifier?: string;
  url?: string;
  username?: string;
  attrs?: any;
}
const Avatar: FunctionComponent<IAvatarProps> = ({
  modifier,
  url,
  username,
  ...attr
}): JSX.Element => {
  const [avatarUrl, setAvatar] = useState<string | undefined>(url);
  const [invalidUrl, setInvalidUrl] = useState<boolean>(true);
  const buildedModifier = modifier ? `avatar--${modifier}` : "";


  const getInitilas = (): string => {
    if (!username) {
      return '!';
    }
    return username
      .split(" ")
      .map(subName => subName[0].toUpperCase())
      .join("")
      .substring(0, 2);
  }

  const initials: string = getInitilas();

  if (url && invalidUrl) {
    const imageData = new Image();
    imageData.src = url;

    imageData.addEventListener('load', (): void => {
      setAvatar(url)
      setInvalidUrl(false);
    });
  }

  return (
    <div className={`avatar ${buildedModifier}`}>
      {avatarUrl ? (
        <img src={avatarUrl} alt="avatar" {...attr} />
      ) : (
          <span>{initials}</span>
        )}
    </div>
  );
};

Avatar.defaultProps = {
  modifier: "",
  url: "",
  username: ""
};
export default Avatar;
