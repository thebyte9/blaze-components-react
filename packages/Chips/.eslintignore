/**/dist/

/node_modules/

/**/node_modules/

*.test.tsx

*.js