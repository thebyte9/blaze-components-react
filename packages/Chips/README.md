## Description

React chips are small blocks of information. They are most commonly used for contacts, tags, selecting an option, filtering content, etc.

## Usage

```js
<Chips
  modifiers={[
    Chips.availableModifiers.parent.deletable,
    Chips.availableModifiers.parent.small
  ]}
>
  <Chips.Avatar>
    <Avatar username="Lorem Ipsum" modifier="x-small" />
  </Chips.Avatar>
  <Chips.Label>Primary deletable chip</Chips.Label>
  <Chips.Icon modifier={Chips.availableModifiers.icon.delete}>
    <i className="material-icons">delete</i>
  </Chips.Icon>
</Chips>
```

## API

##### Chips can receive a number of `props` as follow:

| NAME      |         TYPE          | DEFAULT  |                    OPTIONS                     |
| :-------- | :-------------------: | :------: | :--------------------------------------------: |
| modifiers |    Array of string    | primary  | deletable, outlined, secondary, small, primary |
| action    |       Function        | () => {} |
| children  | single/array of nodes |  empty   |
