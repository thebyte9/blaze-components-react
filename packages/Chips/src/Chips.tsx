import { buildClassNames } from '@blaze-react/utils';
import React, { Fragment, ReactElement, useState } from 'react';
import ChipAvatar from './ChipAvatar';
import ChipIcon from './ChipIcon';
import ChipLabel from './ChipLabel';

type TModifiers = 'deletable' | 'outlined' | 'primary' | 'secondary' | 'small' | string;

interface IChipsProps {
  children: JSX.Element | JSX.Element[];
  modifiers?: TModifiers[];
  action: () => void;
  onDelete: () => void;
}

const Chips = ({ children, modifiers, action, onDelete }: IChipsProps): JSX.Element => {
  const [showChip, setChip] = useState<boolean>(true);

  const deletable: TModifiers = 'deletable';

  const isRemovable: boolean = modifiers && modifiers.includes(deletable);

  const formatedModifiers: string = modifiers ? modifiers.map((modifier) => `chip--${modifier}`).join(' ') : '';

  const handleRemoveChip = (): void => {
    onDelete();
    setChip(false);
  };

  const chipClassNames: string = buildClassNames('chip', {
    [formatedModifiers]: !!modifiers,
  });

  return (
    <Fragment>
      {showChip && (
        <div className={chipClassNames} onClick={action}>
          {isRemovable
            ? React.Children.map(
                children,
                (child: ReactElement): ReactElement =>
                  React.cloneElement(child, {
                    handleRemoveChip,
                  }),
              )
            : children}
        </div>
      )}
    </Fragment>
  );
};

const availableModifiers = {
  icon: {
    custom: 'custom',
    delete: 'delete',
  },
  parent: {
    deletable: 'deletable',
    outlined: 'outlined',
    primary: 'primary',
    secondary: 'secondary',
    small: 'small',
  },
};

Chips.Avatar = ChipAvatar;
Chips.Label = ChipLabel;
Chips.Icon = ChipIcon;

Chips.availableModifiers = availableModifiers;

Chips.defaultProps = {
  action: undefined,
  onDelete: undefined,
};

export default Chips;
