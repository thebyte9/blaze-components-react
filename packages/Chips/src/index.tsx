import withUtils from "@blaze-react/utils";
import React from "react";
import ChipAvatar from "./ChipAvatar";
import ChipIcon from "./ChipIcon";
import ChipLabel from "./ChipLabel";

interface IChipsProps {
  children: JSX.Element | JSX.Element[];
  modifiers: string[];
  utils: {
    classNames: (className: string | object, classNames?: object) => string;
  };
}

const Chips = withUtils(
  ({
    children,
    modifiers,
    utils: { classNames }
  }: IChipsProps): JSX.Element => {
    const chipClassName = classNames("chip", {
      [modifiers
        .map(modifier => `chip--${modifier}`)
        .join(" ")]: !!modifiers.length
    });

    return (
      <div className="chip__wrapper">
        <div className={chipClassName}>{children}</div>
      </div>
    );
  }
);

Chips.Avatar = ChipAvatar;
Chips.Label = ChipLabel;
Chips.Icon = ChipIcon;

export default Chips;
