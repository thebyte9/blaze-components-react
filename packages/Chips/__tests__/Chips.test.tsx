import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import Avatar from '../../Avatar/src';
import Chips from '../src';
import React from 'react';

describe('Chip component', () => {
  test('should be defined and renders correctly (snapshot)', () => {
    const { asFragment } = render(
      <Chips modifiers={[Chips.availableModifiers.parent.deletable, Chips.availableModifiers.parent.small]}>
        <Chips.Label>Basic chip</Chips.Label>
      </Chips>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should delete Chip', () => {
    const { asFragment } = render(
      <Chips
        modifiers={[Chips.availableModifiers.parent.deletable, Chips.availableModifiers.parent.small]}
        action={() => ({})}
      >
        <Chips.Avatar>
          <Avatar username="Lorem Ipsum" modifier="x-small" />
        </Chips.Avatar>
        <Chips.Label>Primary deletable chip</Chips.Label>
        <Chips.Icon modifier={Chips.availableModifiers.icon.delete}>
          <i className="material-icons">delete</i>
        </Chips.Icon>
      </Chips>,
    );

    const deletable = screen.getByText('delete');
    fireEvent.click(deletable);

    expect(asFragment()).toMatchSnapshot();
  });
});
