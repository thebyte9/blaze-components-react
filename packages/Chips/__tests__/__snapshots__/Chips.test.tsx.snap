// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip component should be defined and renders correctly (snapshot) 1`] = `
<withUtils()
  action={[Function]}
  onDelete={[Function]}
>
  <Component
    action={[Function]}
    onDelete={[Function]}
    utils={
      Object {
        "ErrorMessage": [Function],
        "classNames": [Function],
        "removeExtraSpaces": [Function],
        "uniqueId": [Function],
      }
    }
  >
    <div
      className="chip__wrapper"
    >
      <div
        className="chip"
        onClick={[Function]}
      >
        <ChipLabel>
          <div
            className="chip__label"
          >
            Basic chip
          </div>
        </ChipLabel>
      </div>
    </div>
  </Component>
</withUtils()>
`;
